File: examples/language-impl-comparison/haskell/markdown_parser.hs
{-# LANGUAGE OverloadedStrings #-}

-- Markdown風軽量マークアップパーサー - Haskell実装
--
-- Unicode処理の注意点：
-- - HaskellのStringは[Char]でUnicodeコードポイントのリスト
-- - Textはより効率的だがコードポイント単位の操作
-- - Grapheme（書記素クラスター）処理にはtext-icu等の外部ライブラリが必要
-- - Remlの3層モデルと比較すると、Haskellも明示的な区別が必要

module MarkdownParser
  ( Block(..)
  , Inline(..)
  , Document
  , parseMarkdown
  , renderToString
  ) where

import Data.Text (Text)
import qualified Data.Text as T
import Control.Applicative ((<|>))
import Data.Char (isSpace)
import Data.Maybe (fromMaybe)

-- | Markdown AST のブロック要素
data Block
  = Heading { level :: Int, inline :: [Inline] }
  | Paragraph { inline :: [Inline] }
  | UnorderedList { items :: [[Inline]] }
  | OrderedList { items :: [[Inline]] }
  | CodeBlock { lang :: Maybe Text, code :: Text }
  | HorizontalRule
  deriving (Show, Eq)

-- | Markdown AST のインライン要素
data Inline
  = Text Text
  | Strong [Inline]
  | Emphasis [Inline]
  | Code Text
  | Link { text :: [Inline], url :: Text }
  | LineBreak
  deriving (Show, Eq)

type Document = [Block]

-- | パーサー状態
--
-- Unicode処理の課題：
-- - Textはコードポイント単位だが、Grapheme（書記素クラスター）ではない
-- - 「🇯🇵」のような国旗絵文字は2つのコードポイント
-- - 正確な表示幅計算にはtext-icuやunicode-data等が必要
data ParseState = ParseState
  { input :: Text
  , position :: Int  -- コードポイント単位のインデックス（Textの制約）
  }

newtype Parser a = Parser { runParser :: ParseState -> Either String (a, ParseState) }

instance Functor Parser where
  fmap f (Parser p) = Parser $ \s -> case p s of
    Left err -> Left err
    Right (x, s') -> Right (f x, s')

instance Applicative Parser where
  pure x = Parser $ \s -> Right (x, s)
  (Parser pf) <*> (Parser px) = Parser $ \s -> case pf s of
    Left err -> Left err
    Right (f, s') -> case px s' of
      Left err -> Left err
      Right (x, s'') -> Right (f x, s'')

instance Monad Parser where
  return = pure
  (Parser p) >>= f = Parser $ \s -> case p s of
    Left err -> Left err
    Right (x, s') -> runParser (f x) s'

-- | パーサーの選択（左優先）
(<||>) :: Parser a -> Parser a -> Parser a
(Parser p1) <||> (Parser p2) = Parser $ \s -> case p1 s of
  Left _ -> p2 s
  Right res -> Right res

-- | パーサーの失敗
failParser :: String -> Parser a
failParser msg = Parser $ \_ -> Left msg

-- | 現在位置の1文字を取得
--
-- 注意：HaskellのCharはUnicodeコードポイントだが、Graphemeではない
peekChar :: Parser (Maybe Char)
peekChar = Parser $ \s ->
  let remaining = T.drop (position s) (input s)
  in Right (if T.null remaining then Nothing else Just (T.head remaining), s)

-- | 1文字を消費
advanceChar :: Parser ()
advanceChar = Parser $ \s ->
  Right ((), s { position = position s + 1 })

-- | 固定文字列をマッチ
matchString :: Text -> Parser Bool
matchString target = Parser $ \s ->
  let remaining = T.drop (position s) (input s)
  in if T.isPrefixOf target remaining
     then Right (True, s { position = position s + T.length target })
     else Right (False, s)

-- | 水平空白をスキップ
skipHSpace :: Parser ()
skipHSpace = do
  mc <- peekChar
  case mc of
    Just c | c == ' ' || c == '\t' -> advanceChar >> skipHSpace
    _ -> return ()

-- | 行末まで読む
readUntilEol :: Parser Text
readUntilEol = Parser $ \s ->
  let remaining = T.drop (position s) (input s)
      (line, _) = T.break (== '\n') remaining
  in Right (line, s { position = position s + T.length line })

-- | 改行を消費
consumeNewline :: Parser Bool
consumeNewline = do
  mc <- peekChar
  case mc of
    Just '\n' -> advanceChar >> return True
    _ -> return False

-- | EOFチェック
isEof :: Parser Bool
isEof = Parser $ \s ->
  Right (position s >= T.length (input s), s)

-- | 見出し行のパース（`# Heading` 形式）
parseHeading :: Parser Block
parseHeading = do
  skipHSpace

  -- `#` の連続をカウント
  lvl <- countHashes 0

  if lvl == 0 || lvl > 6
    then failParser "見出しレベルは1-6の範囲内である必要があります"
    else do
      skipHSpace
      txt <- readUntilEol
      _ <- consumeNewline
      return $ Heading lvl [Text (T.strip txt)]
  where
    countHashes n = do
      mc <- peekChar
      case mc of
        Just '#' -> advanceChar >> countHashes (n + 1)
        _ -> return n

-- | 水平線のパース（`---`, `***`, `___`）
parseHorizontalRule :: Parser Block
parseHorizontalRule = do
  skipHSpace
  txt <- readUntilEol
  _ <- consumeNewline

  let trimmed = T.strip txt
      isRule = (T.all (== '-') trimmed && T.length trimmed >= 3)
            || (T.all (== '*') trimmed && T.length trimmed >= 3)
            || (T.all (== '_') trimmed && T.length trimmed >= 3)

  if isRule
    then return HorizontalRule
    else failParser "水平線として認識できません"

-- | コードブロックのパース（```言語名）
parseCodeBlock :: Parser Block
parseCodeBlock = do
  matched <- matchString "```"
  if not matched
    then failParser "コードブロック開始が見つかりません"
    else do
      langLine <- readUntilEol
      _ <- consumeNewline

      let language = let trimmed = T.strip langLine
                     in if T.null trimmed then Nothing else Just trimmed

      codeLines <- readCodeLines []
      _ <- consumeNewline

      return $ CodeBlock language (T.intercalate "\n" codeLines)
  where
    readCodeLines acc = do
      matched <- matchString "```"
      if matched
        then return (reverse acc)
        else do
          eof <- isEof
          if eof
            then return (reverse acc)
            else do
              line <- readUntilEol
              _ <- consumeNewline
              readCodeLines (line : acc)

-- | リスト項目のパース（簡易版：`-` または `*`）
parseUnorderedList :: Parser Block
parseUnorderedList = do
  items <- parseItems []
  if null items
    then failParser "リスト項目が見つかりません"
    else return $ UnorderedList items
  where
    parseItems acc = do
      skipHSpace
      mc <- peekChar
      case mc of
        Just c | c == '-' || c == '*' -> do
          advanceChar
          skipHSpace
          txt <- readUntilEol
          _ <- consumeNewline
          parseItems ([Text (T.strip txt)] : acc)
        _ -> return (reverse acc)

-- | 段落のパース（簡易版：空行まで）
parseParagraph :: Parser Block
parseParagraph = do
  lines <- readLines []
  let txt = T.intercalate " " lines
  return $ Paragraph [Text (T.strip txt)]
  where
    readLines acc = do
      eof <- isEof
      if eof
        then return (reverse acc)
        else do
          mc <- peekChar
          case mc of
            Just '\n' -> do
              advanceChar
              mc2 <- peekChar
              case mc2 of
                Just '\n' -> return (reverse acc)
                _ -> readLines ("" : acc)
            _ -> do
              line <- readUntilEol
              _ <- consumeNewline
              readLines (line : acc)

-- | ブロック要素のパース（優先順位付き試行）
parseBlock :: Parser Block
parseBlock = do
  skipBlankLines
  eof <- isEof
  if eof
    then failParser "EOF"
    else do
      skipHSpace
      mc <- peekChar
      case mc of
        Just '#' -> parseHeading
        Just '`' -> do
          matched <- matchString "```"
          if matched
            then parseCodeBlock
            else parseParagraph
        Just c | c == '-' || c == '*' || c == '_' ->
          parseHorizontalRule <||> parseUnorderedList
        _ -> parseParagraph
  where
    skipBlankLines = do
      mc <- peekChar
      case mc of
        Just '\n' -> advanceChar >> skipBlankLines
        _ -> return ()

-- | ドキュメント全体のパース
parseDocument :: Parser Document
parseDocument = parseBlocks []
  where
    parseBlocks acc = do
      result <- (Just <$> parseBlock) <||> return Nothing
      case result of
        Just block -> parseBlocks (block : acc)
        Nothing -> return (reverse acc)

-- | パブリックAPI：文字列からドキュメントをパース
parseMarkdown :: Text -> Either String Document
parseMarkdown input =
  let initialState = ParseState input 0
  in case runParser parseDocument initialState of
    Left err -> Left err
    Right (doc, _) -> Right doc

-- | 簡易的なレンダリング（検証用）
renderToString :: Document -> Text
renderToString doc = T.concat $ map renderBlock doc
  where
    renderInline :: [Inline] -> Text
    renderInline inlines = T.concat $ map renderInlineElem inlines

    renderInlineElem :: Inline -> Text
    renderInlineElem (Text t) = t
    renderInlineElem (Strong inner) = "**" <> renderInline inner <> "**"
    renderInlineElem (Emphasis inner) = "*" <> renderInline inner <> "*"
    renderInlineElem (Code t) = "`" <> t <> "`"
    renderInlineElem (Link txt url) = "[" <> renderInline txt <> "](" <> url <> ")"
    renderInlineElem LineBreak = "\n"

    renderBlock :: Block -> Text
    renderBlock (Heading lvl inlines) =
      T.replicate lvl "#" <> " " <> renderInline inlines <> "\n\n"
    renderBlock (Paragraph inlines) =
      renderInline inlines <> "\n\n"
    renderBlock (UnorderedList items) =
      T.concat (map (\item -> "- " <> renderInline item <> "\n") items) <> "\n"
    renderBlock (OrderedList items) =
      T.concat (zipWith (\i item -> T.pack (show i) <> ". " <> renderInline item <> "\n")
                        [1..] items) <> "\n"
    renderBlock (CodeBlock mLang codeText) =
      let langStr = fromMaybe "" mLang
      in "```" <> langStr <> "\n" <> codeText <> "\n```\n\n"
    renderBlock HorizontalRule = "---\n\n"

-- Unicode 3層モデル比較：
--
-- Haskellでは String = [Char] で Char はUnicodeコードポイント
-- Textはより効率的だが、やはりコードポイント単位の操作
-- Grapheme（書記素クラスター）処理には text-icu 等の外部ライブラリが必要：
--
-- import Data.Text.ICU (normalize, NormalizationMode(..))
-- import Data.Text.ICU.Break (breakCharacter)
--
-- countGraphemes :: Text -> Int
-- countGraphemes = length . breakCharacter (...)
--
-- countCodepoints :: Text -> Int
-- countCodepoints = T.length
--
-- byteLength :: Text -> Int
-- byteLength = BS.length . encodeUtf8
--
-- この明示性の欠如が、絵文字や結合文字の扱いでバグを生む可能性がある。

File: examples/language-impl-comparison/nim/markdown_parser.nim
# Markdown風軽量マークアップパーサー - Nim実装
#
# Unicode処理の注意点：
# - Nimのstringはバイトシーケンス（UTF-8を想定）
# - len() はバイト数を返す
# - runeLen() や runesLen() でUnicodeコードポイント数を取得
# - Grapheme（書記素クラスター）単位の操作は標準ライブラリにない
# - Remlの3層モデル（Byte/Char/Grapheme）と比較すると、Nimも明示的な区別が必要

import std/[strutils, sequtils, options]

# Markdown AST のインライン要素
type
  Inline = ref object
    case kind: InlineKind
    of ikText: text: string
    of ikStrong: strongInner: seq[Inline]
    of ikEmphasis: emphasisInner: seq[Inline]
    of ikCode: code: string
    of ikLink:
      linkText: seq[Inline]
      url: string
    of ikLineBreak: discard

  InlineKind = enum
    ikText, ikStrong, ikEmphasis, ikCode, ikLink, ikLineBreak

# Markdown AST のブロック要素
type
  Block = ref object
    case kind: BlockKind
    of bkHeading:
      level: int
      headingInline: seq[Inline]
    of bkParagraph:
      paragraphInline: seq[Inline]
    of bkUnorderedList:
      ulItems: seq[seq[Inline]]
    of bkOrderedList:
      olItems: seq[seq[Inline]]
    of bkCodeBlock:
      lang: Option[string]
      codeContent: string
    of bkHorizontalRule: discard

  BlockKind = enum
    bkHeading, bkParagraph, bkUnorderedList, bkOrderedList, bkCodeBlock, bkHorizontalRule

  Document = seq[Block]

# パーサー状態
type
  ParseState = object
    input: string
    position: int

  ParseError = object of CatchableError

# コンストラクタヘルパー
proc newText(s: string): Inline =
  Inline(kind: ikText, text: s)

proc newHeading(lvl: int, inline: seq[Inline]): Block =
  Block(kind: bkHeading, level: lvl, headingInline: inline)

proc newParagraph(inline: seq[Inline]): Block =
  Block(kind: bkParagraph, paragraphInline: inline)

proc newUnorderedList(items: seq[seq[Inline]]): Block =
  Block(kind: bkUnorderedList, ulItems: items)

proc newCodeBlock(lang: Option[string], code: string): Block =
  Block(kind: bkCodeBlock, lang: lang, codeContent: code)

proc newHorizontalRule(): Block =
  Block(kind: bkHorizontalRule)

# === パーサーユーティリティ ===

proc peekChar(state: ParseState): Option[char] =
  if state.position >= state.input.len:
    return none(char)
  some(state.input[state.position])

proc advanceChar(state: var ParseState) =
  state.position += 1

proc matchString(state: var ParseState, target: string): bool =
  let remaining = state.input[state.position..^1]
  if remaining.startsWith(target):
    state.position += target.len
    return true
  false

proc skipHSpace(state: var ParseState) =
  while true:
    let ch = peekChar(state)
    if ch.isSome and (ch.get == ' ' or ch.get == '\t'):
      advanceChar(state)
    else:
      break

proc skipBlankLines(state: var ParseState) =
  while true:
    let ch = peekChar(state)
    if ch.isSome and ch.get == '\n':
      advanceChar(state)
    else:
      break

proc readUntilEol(state: var ParseState): string =
  var line = ""
  while true:
    let ch = peekChar(state)
    if ch.isNone or ch.get == '\n':
      break
    line.add(ch.get)
    advanceChar(state)
  line

proc consumeNewline(state: var ParseState) =
  let ch = peekChar(state)
  if ch.isSome and ch.get == '\n':
    advanceChar(state)

proc isEof(state: ParseState): bool =
  state.position >= state.input.len

# === パース関数 ===

proc parseHeading(state: var ParseState): Block =
  skipHSpace(state)

  # `#` の連続をカウント
  var level = 0
  while true:
    let ch = peekChar(state)
    if ch.isSome and ch.get == '#':
      level += 1
      advanceChar(state)
    else:
      break

  if level == 0 or level > 6:
    raise newException(ParseError, "見出しレベルは1-6の範囲内である必要があります")

  skipHSpace(state)
  let text = readUntilEol(state)
  consumeNewline(state)

  let inline = @[newText(text.strip)]
  newHeading(level, inline)

proc parseHorizontalRule(state: var ParseState): Block =
  skipHSpace(state)
  let text = readUntilEol(state)
  consumeNewline(state)

  let trimmed = text.strip
  let isRule =
    (trimmed.allIt(it == '-') and trimmed.len >= 3) or
    (trimmed.allIt(it == '*') and trimmed.len >= 3) or
    (trimmed.allIt(it == '_') and trimmed.len >= 3)

  if not isRule:
    raise newException(ParseError, "水平線として認識できません")

  newHorizontalRule()

proc parseCodeBlock(state: var ParseState): Block =
  if not matchString(state, "```"):
    raise newException(ParseError, "コードブロック開始が見つかりません")

  let langLine = readUntilEol(state)
  consumeNewline(state)

  let lang =
    let trimmed = langLine.strip
    if trimmed.len == 0:
      none(string)
    else:
      some(trimmed)

  # コードブロック内容を ```閉じまで読む
  var codeLines: seq[string] = @[]
  while true:
    if matchString(state, "```"):
      break
    if isEof(state):
      break
    let line = readUntilEol(state)
    consumeNewline(state)
    codeLines.add(line)

  consumeNewline(state)

  let code = codeLines.join("\n")
  newCodeBlock(lang, code)

proc parseUnorderedList(state: var ParseState): Block =
  var items: seq[seq[Inline]] = @[]

  while true:
    skipHSpace(state)
    let ch = peekChar(state)
    if ch.isNone or (ch.get != '-' and ch.get != '*'):
      break

    advanceChar(state)
    skipHSpace(state)
    let text = readUntilEol(state)
    consumeNewline(state)

    let inline = @[newText(text.strip)]
    items.add(inline)

  if items.len == 0:
    raise newException(ParseError, "リスト項目が見つかりません")

  newUnorderedList(items)

proc parseParagraph(state: var ParseState): Block =
  var lines: seq[string] = @[]

  while not isEof(state):
    let ch = peekChar(state)
    if ch.isSome and ch.get == '\n':
      advanceChar(state)
      let ch2 = peekChar(state)
      if ch2.isSome and ch2.get == '\n':
        break  # 空行で段落終了
      lines.add("")
    else:
      let line = readUntilEol(state)
      consumeNewline(state)
      lines.add(line)

  let text = lines.join(" ").strip
  let inline = @[newText(text)]
  newParagraph(inline)

proc parseBlock(state: var ParseState): Block =
  skipBlankLines(state)

  if isEof(state):
    raise newException(ParseError, "EOF")

  skipHSpace(state)

  let ch = peekChar(state)
  if ch.isNone:
    raise newException(ParseError, "EOF")

  case ch.get
  of '#':
    parseHeading(state)
  of '`':
    if matchString(state, "```"):
      parseCodeBlock(state)
    else:
      parseParagraph(state)
  of '-', '*', '_':
    try:
      parseHorizontalRule(state)
    except ParseError:
      # 水平線でなければリストとして解析
      parseUnorderedList(state)
  else:
    parseParagraph(state)

proc parseDocument(state: var ParseState): Document =
  var blocks: seq[Block] = @[]

  while true:
    try:
      let block = parseBlock(state)
      blocks.add(block)
    except ParseError as e:
      if e.msg == "EOF":
        break
      else:
        raise

  blocks

# パブリックAPI：文字列からドキュメントをパース
proc parse*(input: string): Document =
  var state = ParseState(input: input, position: 0)
  parseDocument(state)

# === レンダリング ===

proc renderInline(inlines: seq[Inline]): string =
  result = ""
  for i in inlines:
    case i.kind
    of ikText:
      result.add(i.text)
    of ikStrong:
      result.add("**" & renderInline(i.strongInner) & "**")
    of ikEmphasis:
      result.add("*" & renderInline(i.emphasisInner) & "*")
    of ikCode:
      result.add("`" & i.code & "`")
    of ikLink:
      result.add("[" & renderInline(i.linkText) & "](" & i.url & ")")
    of ikLineBreak:
      result.add("\n")

proc renderBlock(block: Block): string =
  case block.kind
  of bkHeading:
    let prefix = "#".repeat(block.level)
    prefix & " " & renderInline(block.headingInline) & "\n\n"
  of bkParagraph:
    renderInline(block.paragraphInline) & "\n\n"
  of bkUnorderedList:
    var itemsStr = ""
    for item in block.ulItems:
      itemsStr.add("- " & renderInline(item) & "\n")
    itemsStr & "\n"
  of bkOrderedList:
    var itemsStr = ""
    for i, item in block.olItems:
      itemsStr.add($(i + 1) & ". " & renderInline(item) & "\n")
    itemsStr & "\n"
  of bkCodeBlock:
    let langStr = if block.lang.isSome: block.lang.get else: ""
    "```" & langStr & "\n" & block.codeContent & "\n```\n\n"
  of bkHorizontalRule:
    "---\n\n"

proc renderToString*(doc: Document): string =
  result = ""
  for block in doc:
    result.add(renderBlock(block))

# === テスト ===

when isMainModule:
  echo "=== Nim Markdown パーサー ==="

  let markdown = """
# 見出し1

これは段落です。

- リスト1
- リスト2

```nim
echo "Hello"
```

---
"""

  try:
    let doc = parse(markdown)
    echo "パース成功: ", doc.len, " ブロック"
    echo renderToString(doc)
  except ParseError as e:
    echo "パースエラー: ", e.msg

# === Unicode 3層モデル比較 ===
#
# Nimのstringはバイトシーケンス（UTF-8を想定）で：
# - len() はバイト数を返す
# - runesLen() でUnicodeコードポイント数を取得
# - Grapheme（書記素クラスター）単位の操作は標準ライブラリにない
#
# 例：
# let str = "🇯🇵"  # 国旗絵文字（2つのコードポイント、1つのgrapheme）
# str.len  # => 8 (バイト数)
# str.runesLen  # => 2 (コードポイント数)
#
# Remlの3層モデル（Byte/Char/Grapheme）と比較すると、
# Nimも明示的な区別が必要で、標準APIだけでは絵文字や結合文字の扱いが複雑。
#
# === Reml との比較メモ ===
#
# 1. **代数的データ型（ADT）**
#    Nim: object variant (case object) で ADT 風に記述
#    Reml: 型定義で直接 `type Inline = Text(string) | Strong(...) | ...` と記述
#    - Reml の方が構文が簡潔で、パターンマッチがより自然
#
# 2. **パーサーの実装アプローチ**
#    Nim: 手書き再帰下降パーサー、またはNPeg（PEGパーサーコンビネーター）
#    Reml: Core.Parse コンビネーターで宣言的に記述
#    - Reml はパーサーコンビネーターが標準ライブラリとして統合
#
# 3. **エラーハンドリング**
#    Nim: 例外機構（try-except）が主流、Option 型もサポート
#    Reml: Result<T, E> を標準で提供し、? 演算子で簡潔に記述
#    - Reml の方が関数型スタイルに統一
#
# 4. **性能**
#    Nim: C バックエンドにより、非常に高速
#    Reml: 実装次第だが、同等の性能を目指す
#    - Nim は既に成熟した高性能言語